experimental/grpc_attestation/src/attestation.rs

```diff
@@ -0,0 +1,210 @@
+    handler: fn(&[u8]) -> Vec<u8>,
+}
+
+impl RequestHandler {
+    pub fn new(encryptor: AeadEncryptor, handler: fn(&[u8]) -> Vec<u8>) -> Self {
+        Self { encryptor, handler }
+    }
+
+    /// Decrypts a client request, runs [`RequestHandler::handler`] on them and returns an encrypted
+    /// response.
+    pub async fn handle_request(
+        &mut self,
+        receiver: &mut Receiver,
+    ) -> anyhow::Result<Option<GetAttestedDataResponse>> {
+        if let Some(request) = receiver
+            .receive()
+            .await
+            .context("Couldn't receive request")?
+        {
+            let request_type = request.request_type.context("Couldn't read request type")?;
+            let data = if let RequestType::Data(data) = request_type {
+                Ok(data)
+            } else {
+                Err(anyhow!("Received incorrect message type"))
+            }?;
+
+            let decrypted_data = self
+                .encryptor
+                .decrypt(&data)
+                .context("Couldn't decrypt data")?;
+            let response = (self.handler)(&decrypted_data);
+
+            let encrypted_data = self
+                .encryptor
+                .encrypt(&response)
+                .context("Couldn't encrypt data")?;
+            let encrypted_response = GetAttestedDataResponse {
+                response_type: Some(ResponseType::Data(encrypted_data)),
+            };
+
+            Ok(Some(encrypted_response))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// gRPC Attestation Service implementation.
+pub struct AttestationServer {
+    /// PEM encoded X.509 certificate that signs TEE firmware key.
+    tee_certificate: Vec<u8>,
+    /// Processes data from client requests and creates responses.
+    request_handler: fn(&[u8]) -> Vec<u8>,
+}
+
+impl AttestationServer {
+    pub fn create(
+        tee_certificate: Vec<u8>,
+        request_handler: fn(&[u8]) -> Vec<u8>,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            tee_certificate,
+            request_handler,
+        })
+    }
+
+    /// Attest a single gRPC streaming request. Client messages are provided via `receiver`.
+    async fn attest(
+        receiver: &mut Receiver,
+        tee_certificate: Vec<u8>,
+    ) -> anyhow::Result<(GetAttestedDataResponse, AeadEncryptor)> {
+        let request = receiver
+            .receive()
+            .await
+            .context("Couldn't receive attestation request")?
+            .context("Stream stopped preemptively")?;
+
+        // Receive client's attestation request containing a public key.
+        let request_type = request.request_type.context("Couldn't read request type")?;
+        let attestation_request =
+            if let RequestType::AttestationRequest(attestation_request) = request_type {
+                Ok(attestation_request)
+            } else {
+                Err(anyhow!("Received incorrect message type"))
+            }?;
```

Comment:
It would be nice to model the state transitions more cleanly with Rust types at some point. For instance, have a trait for an object that receives messages from a `Receiver`, and returns a new instance of an object that impl that trait, or an error; then we can have one object for each state in the state transition diagram, that consumes itself by value and produces the next step (or an error). I think https://hoverbear.org/blog/rust-state-machine-pattern/ has a good overview, but it has been a while since I last read it, perhaps leave the current code as it is and look into it separately, tracking in a GitHub issue? Please do read that article carefully first, it has a few options, it's not just the first one that it is mentioned.